packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+          neverDrawn: viewport.neverDrawn?.size ? setIntersection(props.elements, viewport.neverDrawn) : undefined,
+          totalCount: props.elements.size,
+        }),
+      );
+    }
+
+    return forkJoin({
+      totalCount: this._queryHandler.queryElementsCount(props.queryProps),
+      alwaysDrawn: this.getAlwaysDrawnElements(props.queryProps),
+      neverDrawn: this.getNeverDrawnElements(props.queryProps),
+    }).pipe(
+      map((state) => {
+        return this.getVisibilityFromAlwaysAndNeverDrawnElementsImpl({
+          ...props,
+          ...state,
+        });
+      }),
+    );
+  }
+
+  private getAlwaysDrawnElements(props: AlwaysOrNeverDrawnElementsQueryProps) {
+    return this._alwaysAndNeverDrawnElements.getElements({ ...props, setType: "always" });
+  }
+
+  private getNeverDrawnElements(props: AlwaysOrNeverDrawnElementsQueryProps) {
+    return this._alwaysAndNeverDrawnElements.getElements({ ...props, setType: "never" });
+  }
+
+  private clearAlwaysAndNeverDrawnElements(props: Omit<AlwaysOrNeverDrawnElementsQueryProps, "setType">) {
+    return forkJoin({
+      alwaysDrawn: this.getAlwaysDrawnElements(props),
+      neverDrawn: this.getNeverDrawnElements(props),
+    }).pipe(
+      map(({ alwaysDrawn, neverDrawn }) => {
+        const viewport = this._props.viewport;
+        if (viewport.alwaysDrawn?.size && alwaysDrawn.size) {
+          viewport.setAlwaysDrawn(setDifference(viewport.alwaysDrawn, alwaysDrawn));
+        }
+        if (viewport.neverDrawn?.size && neverDrawn.size) {
+          viewport.setNeverDrawn(setDifference(viewport.neverDrawn, neverDrawn));
+        }
+      }),
+    );
+  }
+
+  private createVoidOverrideProps<TProps>(props: TProps, obs: Observable<void>): OverridableMethodProps<(props: TProps) => Promise<void>> {
+    return {
+      ...props,
+      originalImplementation: async () => toVoidPromise(obs),
+      handler: this,
+    };
+  }
+
+  private createOverrideProps<TProps, TObservable extends Observable<any>>(
+    props: TProps,
+    obs: TObservable,
+  ): TObservable extends Observable<infer T> ? OverridableMethodProps<(props: TProps) => Promise<T>> : never;
+  private createOverrideProps<TProps, TObservable extends Observable<unknown>>(
+    props: TProps,
+    obs: TObservable,
+  ): OverridableMethodProps<(props: TProps) => Promise<unknown>> {
+    return {
+      ...props,
+      originalImplementation: async () => firstValueFrom(obs),
+      handler: this,
+    };
+  }
+
+  private getGroupedElementIds(node: GroupingHierarchyNode) {
+    const cacheKey = JSON.stringify(node.key);
+    let obs = this._groupingNodeElementsCache.get(cacheKey);
+    if (obs) {
+      return obs;
+    }
+
+    const ids = node.groupedInstanceKeys.map((key) => key.id);
+    const elementIds = from(ids).pipe(
+      mergeWith(this._queryHandler.queryElements({
+        rootElementIds: new Set(ids),
+      })),
+      shareReplay(),
+    );
```

Comment:
When we were talking about element visibility we've come to conclusion that it doesn't make sense for element to be partially visible since it has its own geometry. Class grouping node has no geometry and it's visibility can only be described by its elements, so it makes sense for it to have a hierarchy based state, does it not?

Reply:
yes, but only the grouped elements and not their children

Reply:
Removed querying of recursive children